### ISSUE TITLE
Updates getting dropped after continue as new

### ISSUE THREAD

**Issue by drewhoskins-temporal** [temporalio/features#467]
### Is your feature request related to a problem? Please describe.

It seems continue-as-new can cause near-simultaneous updates to drop, and it's difficult to write the code to figure out what to do about it.

Also, I feel like many users will be caught unawares.  The docs call attention to this problem, which is good, but I feel lucky that I noticed that.

As far as remediating it in my code, the simplest version I thought of was to bookend every update with:

```python
self.num_running_updates += 1
# do stuff
self.num_running_updates -= 1
```

and then use 
```
await workflow.wait_condition(self.running_updates==0)
```

at the end.  But this was not easy to work out and feels "advanced."

### Describe the solution you'd like

I don't have a specific solution in mind, but I'd ideally like the SDK to prevent this from happening, or at least inform me that it is happening and provide a facility for me to track the outstanding updates so I can wait for them.

### Per-SDK Tickets

- [ ] Go - 
- [ ] Java - 
- [ ] Core - 
- [ ] TypeScript - 
- [ ] Python - 
- [ ] .NET - 
- [ ] PHP - 
- [ ] Temporal CLI -

**Comment by Quinn-With-Two-Ns** [comment]
Note the same applies to completing a workflow (since that is what continue as new is). When you complete a workflow all in-progress updates are dropped. I don't think we should build anything special for **just** continue as new. ie if we want to have some warning it should apply anytime a users tries to complete a workflow with non-complete updates,

**Comment by cretz** [comment]
My preference: completing a workflow run today (via failure w/ retry, continue-as-new, etc) should consistently not carryover anything (signals, in-flight queries, in-flight updates, activities, child workflows, etc). We can have a general discussion about carryover on everything at some point though.

For pending admitted-only updates, we should reject workflow completion attempts with unhandled-command same as we do for signals. For accepted updates that are not completed by the workflow, we should fail the update as workflow completed.

**Comment by drewhoskins-temporal** [comment]
@cretz I think it's too early to shut this conversation down.  I do agree we should consider other similar cases as we design it, but there's no reason update can't lead the way.  Update is also quite distinct from those other cases and enough different that impacts of shipping as-is should be considered separately.

**Comment by drewhoskins-temporal** [comment]
I don't agree that consistency trumps other considerations here.  Consistency is one of several concerns to weigh.  
I also wouldn't call the current approach simple--it's simple for us, but could be complex and mind-bending for the customers who encounter it.

**Comment by cretz** [comment]
It's not that consistent trumps other things, it's that we have a chance to consistently solve for updates in the near term, and consistently solve for everything in the long term. I don't believe it's worth giving update special continue-as-new behavior in the near term is valuable (in fact, if we care about this type value, we should do it with activities and child workflows immediately where there is arguably more value than update).

**Comment by drewhoskins-temporal** [comment]
Agreed that continue-as-new carryover is a larger discussion.  But I'm not ready to concede that we should do nothing now in hopes that it will happen one day.  I don't like to bank on things unless they're actually on the immediate roadmap.

In theory, you're right that activity is much more common.  But activities are usually kicked off by the Workflow and workflows are generally written in such a way that you hold onto a future with the activity and can await them before exiting/continuing.  Whereas signals and updates come from outside and are often handled out of the main workflow flow, making it both (a) much more likely (as a percentage) that they will run into trouble, and (b) harder to figure out what to do about it when it does happen.

All that's to say, the use cases are distinct and there may be strong and general mitigations we can do for the update/signal cases.

**Comment by drewhoskins-temporal** [comment]
I want to raise the idea that the root cause here is the same as that for interleaved signals/updates: we're not serializing, which means it can race with all the things: other signals/updates, workflow code, or the end of the workflow.

**Comment by drewhoskins-temporal** [comment]
This was partially fixed by the addition of mutexes.
